feat(blog): add release notes for 2.4 release (#634)
asyncapi/website: pages/blog/release-notes-2.4.0.md

## Server Variables can be now referenced from components

To allow for more flexibility in how AsyncAPI documents are structured and enable content to be reused, `serverVariables` can now be defined as reusable components.

For example:

```yaml
asyncapi: 2.4.0
servers:
  development:
    $ref: '#/components/servers/myserver'
  production:
    $ref: '#/components/servers/myserver'
components:
  servers:
    myserver:
      url: "{stage}.my-server.com:{port}"
      protocol: ws
      variables:
        stage:
          $ref: "#/components/serverVariables/stage"
        port:
          $ref: "#/components/serverVariables/port"
  serverVariables:
    stage:
      default: dev
    port:
      enum: [5000, 6000]
      default: 5000
```

These are added to the many other aspects of the AsyncAPI specification which can be declared as reusable components. You can see the full list in the [Components Object section of the AsyncAPI specification](https://www.asyncapi.com/docs/specifications/v2.4.0#componentsObject).

> This new feature was contributed by [Daniel Kocot](https://github.com/danielkocot). For more detail, see this [pull request](https://github.com/asyncapi/spec/pull/717) and the [Github issue where this change was discussed](https://github.com/asyncapi/spec/issues/707).

Reviewer: ```suggestion
This new feature was contributed by [Daniel Kocot](https://github.com/danielkocot). For more detail, see [Daniel's `/spec #717` pull request](https://github.com/asyncapi/spec/pull/717) and the [Github issue where Daniel's change to `serverVariables ` was discussed](https://github.com/asyncapi/spec/issues/707).
```

same [accessibility note as below in the other comments](https://github.com/asyncapi/website/pull/634/files#r856601171), just an idea of how to fix it

> ## Security can now be defined at Operation level

Author: @sekharbans-ebay would you like to review the release notes I wrote here about your recent contribution? Also check the example in detail, please.
Thanks!

For example:

```yaml
asyncapi: 2.4.0
servers:
  production:
    url: "mykafkacluster.org:8092"
    protocol: kafka-secure
    security:
      - service_auth:
         - auth:write
         - auth:read
channels:
  some/events:
    servers:
      - production
    subscribe:
      # This operation level security implies the ability to send a message to 
      # `some/events` with Authorization headers that have `auth:write` scope. 
      # Note that the operation level security must still satisfy security options 
      # specified at the server level.
      security: 
       service_auth:
         - auth:read  
```

## Reusability of Servers defined in Components is clarified in the specification

Reusability of Servers was introduced in [AsyncAPI 2.3.0](https://www.asyncapi.com/blog/release-notes-2.3.0#servers-and-channels-can-now-be-defined-as-reusable-components). However, the change was not fully clarified in the specification, leading to confusion.

Thanks to [Vladimir Gorej](https://github.com/char0n), this is now clarified by mentioning that elements for the Servers Object can be either Server Object or a Reference Object.

> For more detail, see this [pull request](https://github.com/asyncapi/spec/pull/706) and the [Github issue where this change was discussed](https://github.com/asyncapi/spec/issues/705).

Reviewer: ```suggestion
For more detail, see [Vladimir's `/spec #706`pull request](https://github.com/asyncapi/spec/pull/706) and the [Github issue where Vladimir's Servers Object change was discussed](https://github.com/asyncapi/spec/issues/705).
```

same [accessibility note as below in the other comments](https://github.com/asyncapi/website/pull/634/files#r856601171), just an idea of how to fix it

## Tooling support

The following official AsyncAPI tools are already updated to support 2.4.0 version of the specification:

> - JSON Schema that supports validation of AsyncAPI documents is updated in [this](https://github.com/asyncapi/spec-json-schemas) repository. Also **@asyncapi/specs** package has been updated on NPM to version `TBD`, and it contains the 2.4.0 JSON Schema.

Author: TBD because uncertainty about https://github.com/asyncapi/spec-json-schemas/pull/184

Reviewer: Should you add that note to the blog post? I think it is confusing to read your blog post otherwise 🤔

Author: Oh yeah, this TBD means I need to change it before publishing the blog post.

- [JavaScript Parser](https://github.com/asyncapi/parser-js/) uses latest **@asyncapi/specs** package and can be used to parse and validate 2.4.0 documents. Upgrade to `v2.14.0` version.
- [HTML template](https://github.com/asyncapi/html-template) uses the latest **@asyncapi/react-component** package. Upgrade to `TBD` version.
- [JavaScript Converter](https://github.com/asyncapi/converter-js/) enables conversion from any AsyncAPI version into the 2.4.0 version of the spec. Upgrade to `TBD` version.
- [Modelina](https://github.com/asyncapi/modelina/) now also accepts AsyncAPI documents valid against the 2.4.0 version of the spec. Upgrade to `TBD` version.
 - [Generator](https://github.com/asyncapi/generator/) uses the latest @asyncapi/parser package, so while generating output, it can validate 2.4.0 documents. Upgrade to `TBD` version.

> Last but not least is the AsyncAPI Studio. Check new studio with [this example](https://studio.asyncapi.com/?url=https://raw.githubusercontent.com/asyncapi/spec/v2.4.0/examples/websocket-gemini.yml).

Reviewer: ```suggestion
Last but not least is the AsyncAPI Studio. [Check out the new Studio with this example](https://studio.asyncapi.com/?url=https://raw.githubusercontent.com/asyncapi/spec/v2.4.0/examples/websocket-gemini.yml).
```

For accessibility, all hyperlinks should include non-generic text. This is because a user depending on screenreader technology will not be able to tell where the link takes them unless they read the text surrounding the link.

- **wrong**: [this pull request](some.url),  [read more](some.url),  [click here](some.url)
- **right**:  [AsyncAPI Studio example](some.url),  [read more about GSoD](some.url),  [click on Sergio's awesome blog post](some.url)

    byline: Pilot at AsyncAPI Airlines
excerpt: 'AsyncAPI 2.4 is now released. This brings really helpful additions, such as the new `messageId' field, Server Variables reusability, and security at Operation level'
featured: true
---

The new version of the AsyncAPI specification - 2.4.0 - is now available.

> This is a minor release, and it doesn't bring any breaking changes. You can switch to it by modifying the following value in your AsyncAPI file `asyncapi: '2.3.0'` into `asyncapi: '2.4.0'`

## Message uniqueness thanks to the new messageId field

As operations have `OperationId`, AsyncAPI messages can now define `messageId`. This new field is used to identify a message across a whole AsyncAPI document uniquely.
This new field will be helpful in tools that ask the user to select a message as input, such as validating their schema or filtering code to be generated.

For example:

```yaml
asyncapi: 2.4.0
components:
  messages:
    SomeMessage:
      messageId: SomeMessage
      payload:
        type: object
        properties:
          name:
            type: string
```

This new feature was contributed by [Waleed Ashraf](https://github.com/WaleedAshraf).

> For more detail, see [Waleed's `/spec #751`pull request](https://github.com/asyncapi/spec/pull/751) and the [Github issue where Waleed's `MessageId` feature addition was discussed](https://github.com/asyncapi/spec/issues/458).

Author: For some reason, I dropped the notes for this feature. Now I recovered them.

cc @WaleedAshraf, in case you want to give an eye.
cc @derberg @alequetzalli as this will be new for you to review.

Reviewer 2: @smoya LGTM 🌮

---
title: AsyncAPI Spec 2.4.0 Release Notes
date: 2022-04-25T19:00:00+01:00
type: Communication
tags:
  - Specification
  - Release Notes
cover: /img/posts/release-notes-2.4.0/cover.webp
authors:
  - name: Sergio Moya
    photo: /img/avatars/smoya.webp
    link: https://twitter.com/smoyac
    byline: Pilot at AsyncAPI Airlines

> excerpt: 'AsyncAPI 2.4 is now released. This brings really helpful additions, such as the new `messageId' field, Server Variables reusability, and security at Operation level'

Reviewer 3: ```suggestion
excerpt: 'AsyncAPI 2.4 is now released. This brings really helpful additions, such as the new `messageId` field, Server Variables reusability, and security at Operation level'
```

## Security can now be defined at Operation level

Until today, Security requirements were defined at Server level. That restricted the security requirements to be the same for all channels linked with a Server, and for all operations of those channels.
In fact, when setting Security in both the Server and the Operation, both should be satisfied.

For example:

```yaml
asyncapi: 2.4.0
servers:
  production:
    url: "mykafkacluster.org:8092"
    protocol: kafka-secure
    security:
      - service_auth:
         - auth:write
         - auth:read
channels:
  some/events:
    servers:
      - production
    subscribe:
      # This operation level security implies the ability to subscribe to messages from
      # `some/events` channel with Authorization headers 
      # that have `auth:read` scope. Note that an operation level security must still satisfy 
      # security requirements specified at the server level.
      security:
        - service_auth:
          - auth:read  
```

> Thanks to [Sekharbans](https://github.com/sekharbans-ebay), is now possible to increase security granularity by defining a set of security requirements at Operation level. For more detail, see [Sekharban's `/spec #584` pull request](https://github.com/asyncapi/spec/pull/584) and the [Github issue where Sekharban's suggested feature was discussed](https://github.com/asyncapi/spec/issues/584).

Author: @alequetzalli I added those two lines as I missed adding the link to the PR and Issue on this feature.
Would you mind giving a check and +1 if you agree please?

Reviewer: why limit to a +1 when I can do an `APPROVE PR STAT` 😸

---
title: AsyncAPI Spec 2.4.0 Release Notes
date: 2022-04-25T19:00:00+01:00
type: Communication
tags:
  - Specification
  - Release Notes
cover: /img/posts/release-notes-2.4.0/cover.webp
authors:
  - name: Sergio Moya
    photo: /img/avatars/smoya.webp
    link: https://twitter.com/smoyac
    byline: Pilot at AsyncAPI Airlines
excerpt: 'AsyncAPI 2.4 is now released. This brings ... TBD ...'
featured: true
---

The new version of the AsyncAPI specification - 2.4.0 - is now available.

> This is a minor release, and it doesn't bring any breaking changes. You can switch to it by modifying the following value in your AsyncAPI file `asyncapi: '2.3.0'` into `asyncapi: '2.4.0'`

> ## Server Variables can be now referenced from components

Author: @danielkocot would you like to review the release notes I wrote here about your recent contribution? Does the example look ok? Thanks!

Reviewer 4: Review done. The example looks right.

Author: Thanks!

For example:

```yaml
asyncapi: 2.4.0
servers:
  development:
    $ref: '#/components/servers/myserver'
  production:
    $ref: '#/components/servers/myserver'
components:
  servers:
    myserver:
      url: "{stage}.my-server.com:{port}"
      protocol: ws
      variables:
        stage:
          $ref: "#/components/serverVariables/stage"
        port:
          $ref: "#/components/serverVariables/port"
  serverVariables:
    stage:
      default: dev
    port:
      enum: [5000, 6000]
      default: 5000
```

These are added to the many other aspects of the AsyncAPI specification which can be declared as reusable components. You can see the full list in the [Components Object section of the AsyncAPI specification](https://www.asyncapi.com/docs/specifications/v2.4.0#componentsObject).

This new feature was contributed by [Daniel Kocot](https://github.com/danielkocot). For more detail, see this [pull request](https://github.com/asyncapi/spec/pull/717) and the [Github issue where this change was discussed](https://github.com/asyncapi/spec/issues/707).

## Security can now be defined at Operation level

> TBD

Author: @sekharbans-ebay I'm about to write the release notes for https://github.com/asyncapi/spec/pull/584. Just in case you have something written down or prepared, so you can share it with me. Otherwise, I will start baking them tomorrow morning (CEST).

> ## Reusability of Servers defined in Components is clarified in the specification

Author: @char0n would you like to review the release notes I wrote here about your recent contribution? Thanks!

Reviewer 5: Sure, I'll look into it during tonight. Thanks for notice.

Author: @char0n thanks!

Reusability of Servers was introduced in [AsyncAPI 2.3.0](https://www.asyncapi.com/blog/release-notes-2.3.0#servers-and-channels-can-now-be-defined-as-reusable-components). However, the change was not fully clarified in the specification, leading to confusion.

> Thanks to [Vladimir Gorej](https://github.com/char0n), this is now clarified by mentioning that elements for the Servers Object can be either Server Object or a Reference Object.

Reviewer 5: ```suggestion
Thanks to [Vladimir Gorej](https://github.com/char0n), this is now clarified by mentioning that elements for the Servers Object can be either a Server Object or a Reference Object.
```
